File: php/hilo.php
<?php
function hilo() {
    while (true) {
        echo "\nWelcome to Hi-Lo! The rules are simple: you have 7 tries to guess a number 1-100.
If your guess is too low, I'll say 'Too low'; if your guess is too high, I'll say 'Too high'
If you guess correctly, you win!\n\n";

        $answer = rand(1, 100);
        $guess = 0;

        for ($attempt = 1; $attempt <= 7; $attempt += 1) {
            while (true) {
                echo "Guess $attempt (1-100): ";
                try {
                    $guess = (int)readline();
                    if ($guess >= 1 and $guess <= 100)
                        break;
                }
                catch (exception $e) {}
            }

            if ($guess < $answer)
                echo "Too low.\n";
            elseif ($guess > $answer)
                echo "Too high.\n";
            else {
                echo "Correct! You win!\n";
                break;
            }
        }

        if ($guess != $answer)
            echo "Sorry, you lost. The correct number was $answer.\n";

        echo "\nWould you like to play again? (y/n): ";
        if (readline() != "y") {
            echo "Thanks for playing!\n";
            break;
        }
    }
}

hilo();
?>

File: lua/hilo.lua
function hilo()
    math.randomseed(os.time())

    while true do
        io.write("\nWelcome to Hi-Lo! The rules are simple: you have 7 tries to guess a number 1-100.\n",
                 "If your guess is too low, I'll say \'Too low\'; if your guess is too high, I'll say \'Too high\'\n",
                 "If you guess correctly, you win!\n\n")

        answer = math.random(1, 100)
        guess = 0

        for attempt = 1, 7, 1 do
            while true do
                io.write(string.format("Guess #%s (1-100): ", attempt))
                input = tonumber(io.read())
                if input ~= nil then
                    guess = input
                    if guess >= 1 and guess <= 100 then
                        break
                    end
                end
            end

            if guess < answer then
                io.write("Too low.\n")
            elseif guess > answer then
                io.write("Too high.\n")
            else
                io.write("Correct! You win!\n")
                break
            end
        end

        if guess ~= answer then
            io.write("Sorry, you lost. The correct number was {answer}.\n")
        end

        io.write("\nWould you like to play again? (y/n): ")
        if io.read() ~= "y" then
            io.write("Thanks for playing!\n")
            break
        end
    end
end

hilo()

